File: codeforces/1900~1999/1968/g2.cc
#include <bits/stdc++.h>

using namespace std;

using i64 = long long;

vector<int> z_function(const string& s) {
  int n = s.length();
  vector<int> z(n);
  int l = 0, r = 0;
  for (int i = 1; i < n; ++i) {
    if (i < r) {
      z[i] = min(r-i, z[i-l]);
    }
    while (i + z[i] < n && s[z[i]] == s[i+z[i]]) {
      z[i]++;
    }
    if (i + z[i] > r) {
      l = i;
      r = i + z[i];
    }
  }
  return z;
}

void solve() {
  int n = 0, l = 0, r = 0;
  cin >> n >> l >> r;
  string s;
  cin >> s;

  vector<int> z = z_function(s);

  auto possible = [&](int k, int len) -> bool {
    int count = 1;
    for (int i = len; i < n;) {
      if (z[i] >= len) {
        count++;
        i += len;
      } else {
        i++;
      }
    }
    return count >= k;
  };

  auto find_max_len = [&](int k) -> int {
    int l = 1, r = n/k;
    int ans = 0;
    while (l <= r) {
      int mid = l + (r-l) / 2;
      if (possible(k, mid)) {
        ans = mid;
        l = mid+1;
      } else {
        r = mid-1;
      }
    }
    return ans;
  };

  int m = sqrt(n);
  vector<int> ans(n+1);
  for (int k = 1; k <= m; ++k) {
    ans[k] = find_max_len(k);
  }

  // for (int i = 0; i < n; ++i) cout << z[i] << " ";
  // cout << endl;

  for (int len = 1; len <= m; ++len) {
    int k = 1;
    for (int i = len; i < n;) {
      if (z[i] >= len) {
        k++;
        i += len;
      } else {
        i++;
      }
    }
    ans[k] = max(ans[k], len);
  }

  for (int i = n-1; i >= 1; --i) {
    ans[i] = max(ans[i], ans[i+1]);
  }

  for (int i = l; i <= r; ++i) {
    cout << ans[i] << " ";
  }
  cout << endl;
}

int main(void) {
  ios::sync_with_stdio(false);
  cin.tie(0);

  int T = 0;
  cin >> T;
  for (int t = 0; t < T; ++t) {
    solve();
  }
  return 0;
}

File: codeforces/1900~1999/1968/g2.rs
use std::io::{stdin, stdout, BufWriter, Write};
use std::str::FromStr;
use std::fmt::Debug;
use std::cmp::min;
use std::cmp::max;

fn read_int<T: FromStr>() -> T where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    s.trim().parse::<T>().expect("invalid input")
}

fn read_vec<T: FromStr>() -> Vec<T> where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    if s.is_empty() {
        return Vec::new();
    }
    s.split(" ").map(|x| x.trim().parse().expect("invalid input")).collect()
}

fn read_str() -> String {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    s.trim().to_string()
}

fn z_function(s: &Vec<char>) -> Vec<usize> {
    let n = s.len();
    let mut l = 0;
    let mut r = 0;
    let mut z = vec![0; n];
    for i in 1..n {
        if i < r {
            z[i] = min(r - i, z[i - l]);
        }
        while i + z[i] < n && s[z[i]] == s[i + z[i]] {
            z[i] += 1;
        }
        if i + z[i] > r {
            l = i;
            r = i + z[i];
        }
    }
    z
}

fn possible(z: &Vec<usize>, len: usize, k: usize) -> bool {
    let n = z.len();
    let mut i = len;
    let mut count = 1;
    while i < n {
        if z[i] >= len {
            i += len;
            count += 1;
        } else {
            i += 1;
        }
    }
    count >= k
}

fn solve() {
    let stdout = stdout();
    let mut writer = BufWriter::new(stdout.lock());
    let inputs = read_vec::<usize>();
    let n = inputs[0];
    let left = inputs[1];
    let right = inputs[2];
    let s: Vec<char> = read_str().chars().collect();

    let mut ans = vec![0; n+1];
    let z = z_function(&s);
    let sq = (n as f32).sqrt() as usize;
    for k in 1..=sq {
        let mut l = 1;
        let mut r = n / k;
        let mut result = 0;
        while l <= r {
            let mid = l + (r-l) / 2;
            if possible(&z, mid as usize, k) {
                l = mid + 1;
                result = mid;
            } else {
                r = mid - 1;
            }
        }
        ans[k] = result;
    }

    for len in 1..=sq {
        let mut k = 1;
        let mut i = len;
        while i < n {
            if z[i] >= len {
                k += 1;
                i += len
            } else {
                i += 1;
            }
        }
        ans[k] = max(ans[k], len);
    }

    for i in (1..n).rev() {
        ans[i] = max(ans[i], ans[i+1]);
    }

    for i in left..=right {
        write!(writer, "{} ", ans[i]).ok();
    }
    writeln!(writer).ok();
}

fn main() {
    let n = read_int::<i32>();
    for _t in 0..n {
        solve();
    }
}
